### ISSUE TITLE
Treat `Mapping` as immutable

### ISSUE THREAD

**Issue by giovannipcarvalho** [astral-sh/ty#3643]
### Summary

According to [1] Mapping does not implement .pop() or assigment, only MutableMapping does so.
This is also how mypy treats it.

```python
from collections.abc import Mapping

m: Mapping[str, int] = {"a": 1, "b": 2}


# mypy: error: Unsupported target for indexed assignment ("Mapping[str, int]")  [index]
# ty: ok
m["c"] = 3

# mypy: error: "Mapping[str, int]" has no attribute "pop"  [attr-defined]
# ty: ok
m.pop("a")
```

[1] https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

### Version

ty 0.0.42

**Comment by carljm** [comment]
Thanks for the report!

We do treat `Mapping` as immutable, in the same way as mypy (which is not surprising, since that is effectively defined by the stubs in typeshed, which all type checkers share.) You can see that behavior in this playground example: https://play.ty.dev/8ea96798-54ea-42b1-8d3d-5c22cfceec1b

The difference in your example is not related to the treatment of `Mapping`, but is a difference with mypy in how type annotations on an annotated assignment are handled. You can [see this difference](https://play.ty.dev/1e038565-ca60-4c73-b933-30ee1e206d97) if you add a `reveal_type(m)` to your example: you will see the revealed type as `dict[str, int]`, not `Mapping[str, int]`. This is a valid inference, because `dict[str, int]` is a subtype of `Mapping[str, int]`, so it doesn't violate the declared type, and ty can clearly see that the actual value of `m` is a dict.

There is an inherent ambiguity in Python as to the meaning of an annotated assignment: is it a cast, meaning that the inferred type of the RHS should always be ignored in favor of the annotated type? Or is it simply declaring an upper bound on the type of the variable? The latter interpretation is suggested by the fact that all type checkers, including mypy, will [happily narrow the type of `m` to a more precise type from a later assignment](https://mypy-play.net/?gist=d33b5eec930b9c7498baebe9e337a1f2). So mypy ignores the more-precise inferred RHS type if it's in the same statement as the annotation, but honors it otherwise.

We discussed these tradeoffs extensively in https://github.com/astral-sh/ty/issues/136. We're aware that the one use case not served well by our current behavior is the one where you want to annotate a variable in order to "upcast" it (e.g. prevent mutation of a dict in the local scope.) If it's really important to have ty treat something as a `Mapping` even when it is clearly always a `dict` at runtime, this requires a `cast`.
